portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,56 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
 import java.io.IOException;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
+import java.util.ArrayList;
+import java.util.List;
 
 /** Servlet that returns some example content. TODO: modify this file to handle comments data */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
+  static final String CONTENT_TYPE = "application/json;";
+
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello Anthony!</h1>");
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+    List messages = new ArrayList<String>();
+    for (Entity entity : results.asIterable()){
+      String content = (String) entity.getProperty("content");
+      messages.add(content);
+    }
+    Gson gson = new Gson();
+    String json = gson.toJson(messages);
+    response.setContentType(CONTENT_TYPE);
+    response.getWriter().println(json);
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String content = request.getParameter("content").trim();
+    long timestamp = System.currentTimeMillis();
+    if (content.length() == 0) {
+      System.err.println("Empty comment submitted");
+      return;
```

Comment:
What happens if you don't specify a response? Would the user be redirected to a blank page instead?

Consider adding a default HttpServletResponse.sendError, or you may (consider it a separate feature for error handling) redirect user to a customized error handling page, such as this one: https://github.com/thisgoesnowhere

Reply:
So I ended up realizing this later in my coding journey. In this PR, they are directed to a blank page, but commit e0f906f3feb09751c94b813a9a4b6ed8c433f860 will redirect the user no matter what. Is there a way to let the user know that they were unsuccessful, while still remaining at the comment box?

Reply:
Great question.

"know that they were unsuccessful, while still remaining at the comment box"

Form submission *is* a redirection so user will land on whatever you response.redirect(...) them to. In order to keep the page from being navigated, you need to send the form in JavaScript. AJAX is the basic format of how this is done. jQuery is a library to simplify this process:
https://api.jquery.com/jquery.ajax/

It is easy to ramp up, and may lend great help for the Capstone project as well.